src/tsickle.ts

```diff
@@ -105,89 +105,177 @@ class ClosureRewriter extends Rewriter {
+                paramData.names[i] = [newTag.parameterName];
+              } else if (paramData.names[i].indexOf(newTag.parameterName) === -1) {
+                paramData.names[i].push(newTag.parameterName);
+              }
+            }
+            // merget type names at this index if required
+            if (newTag.type) {
+              if (!paramData.types[i]) {
+                paramData.types[i] = [newTag.type];
+              } else if (paramData.types[i].indexOf(newTag.type) === -1) {
+                paramData.types[i].push(newTag.type);
+              }
+            }
+          }
+        }
+      }
+      // If this method is not overloaded, we set the list of names to those in the only declaration
+      if (!isOverloaded) {
+        paramNamesToReturn = fnDecl.parameters.map(p => p.name.getText());
+      }
 
-        // Search for this parameter in the JSDoc @params.
-        for (let {tagName, parameterName, text} of jsDoc) {
-          if (tagName === 'param' && parameterName === newTag.parameterName) {
-            newTag.text = text;
-            break;
+      // Return type.
+      if (!isConstructor) {
+        let retType = typeChecker.getReturnTypeOfSignature(sig);
+        let retTypeString: string = this.typeToClosure(fnDecl, retType);
+        if (!isOverloaded) {
+          // TODO: Consider adding text from existing JSDoc into overloads
+          let returnDoc: string|undefined;
+          for (let {tagName, text} of jsDoc) {
+            if (tagName === 'return') {
+              returnDoc = text;
+              break;
+            }
+          }
+          newDoc.push({
+            tagName: 'return',
+            type: retTypeString,
+            text: returnDoc,
+          });
+        } else {
+          if (retTypeString) {
+            if (!paramData.returns) {
+              paramData.returns = retTypeString;
+            } else if (paramData.returns.indexOf(retTypeString) === -1) {
+              paramData.returns += '|' + retTypeString;
+            }
           }
         }
-        newDoc.push(newTag);
       }
     }
-
-    // Return type.
-    if (fnDecl.kind !== ts.SyntaxKind.Constructor) {
-      let retType = typeChecker.getReturnTypeOfSignature(sig);
-      let returnDoc: string|undefined;
-      for (let {tagName, text} of jsDoc) {
-        if (tagName === 'return') {
-          returnDoc = text;
-          break;
+    if (isOverloaded) {
+      // Build actual JSDoc tags for the merged param names/types and return types
+      for (let i = 0; i < maxCount; i++) {
+        paramNamesToReturn.push(paramData.names[i].join('_or_'));
+        if (i >= minCount) {
+          paramNamesToReturn[i] = 'opt_' + paramNamesToReturn[i];
         }
+        let concatenatedTypes = paramData.types[i].join('|');
+        newDoc.push({
+          tagName: 'param',
+          parameterName: paramNamesToReturn[i],
+          type: concatenatedTypes,
+        });
+      }
+      if (!isConstructor) {
```

Comment:
I don't think so. The first is within the for(..) over the different declarations, to build the set of unique list of return types. This is the final "do we emit a tag for @return or not'.